### PR TITLE
[Python] Add `read_csv` method

### PR DESCRIPTION
**Pull request by Tishj**
This method is made to be a copy of the `read_csv` method of Pandas, as far as our current options allow it to be.

The supported options are:
- `header` - can be given as `0`, to be compatible with pandas, or as a boolean
- `dtype` - can be given as a dict{str : str} or as a list(str)
- `sep`|`delimiter` - given as string
- `na_values` - define the null string
- `skiprows` - skip the first n rows
- `compression` - given as string
- `quotechar` - given as string
- `escapechar` given as string
- `encoding` given as string (only utf-8 is supported)

Future improvements:
Add `names`, then we can also support `prefix` by combining it into the `names`
Add `usecols`, this can be done by pushing a projection on top of the read_csv_relation

### CODE REVIEW

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM.

> encoding given as string (only utf-8 is supported)

Unrelated to this PR - but I wonder if we could easily support different encodings using #5829

File: tools/pythonpkg/duckdb_python.cpp [hunk 1]
```diff
@@ -103,6 +103,8 @@ static void InitializeConnectionMethods(py::module_ &m) {
 	         py::arg("connection") = py::none())
 	    .def("rollback", &PyConnectionWrapper::Rollback, "Roll back changes performed within a transaction",
 	         py::arg("connection") = py::none())
+	    .def("read_csv", &PyConnectionWrapper::ReadCSV, "Read the CSV file identifier by 'name'", py::arg("name"),
```
Comment (Mause) [inline tools/pythonpkg/duckdb_python.cpp:106]:
```suggestion
	    .def("read_csv", &PyConnectionWrapper::ReadCSV, "Read the CSV file identified by 'name'", py::arg("name"),
```

File: tools/pythonpkg/src/pyconnection.cpp [hunk 2]
```diff
@@ -402,6 +415,154 @@ shared_ptr<DuckDBPyConnection> DuckDBPyConnection::RegisterPythonObject(const st
...
+
+		bool header_as_int = py::isinstance<py::int_>(header_arg);
+		bool header_as_bool = py::isinstance<py::bool_>(header_arg);
+
+		if (!header_as_int && !header_as_bool) {
+			throw InvalidInputException("read_csv only accepts 'header' as an integer, or a boolean");
+		}
+
+		if (header_as_bool) {
+			if (py::str(header_arg).is(py::str(Py_True))) {
```
Comment (Mause) [inline tools/pythonpkg/src/pyconnection.cpp:438]: This should probably be `if (py::bool_(header_arg))` instead?

File: tools/pythonpkg/src/pyconnection.cpp [hunk 3]
```diff
@@ -402,6 +415,154 @@ shared_ptr<DuckDBPyConnection> DuckDBPyConnection::RegisterPythonObject(const st
 	return shared_from_this();
 }
 
+unique_ptr<DuckDBPyRelation> DuckDBPyConnection::ReadCSV(const string &name, const py::kwargs &kwargs) {
```
Comment (Mause) [inline tools/pythonpkg/src/pyconnection.cpp:418]: Whats the reason you've implemented this using kwargs? Most of the type checking code could be deleted if you used explicitly stated arguments instead

File: tools/pythonpkg/duckdb_python.cpp [hunk 4]
```diff
@@ -103,6 +103,8 @@ static void InitializeConnectionMethods(py::module_ &m) {
...
+	         py::arg("connection") = py::none(), py::kw_only {})
```
Comment (Mause) [inline tools/pythonpkg/duckdb_python.cpp:107]: If you're going to use `kwargs` here, you should list all the supported arguments in the docstring (preferably using the RestructedText Sphinx format)

**Review by Mause — CHANGES_REQUESTED** [review]

File: tools/pythonpkg/duckdb_python.cpp (hunk 4 above)
Comment (Mause) [reply tools/pythonpkg/duckdb_python.cpp:107]: Some reference docs for the syntax: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html

File: tools/pythonpkg/src/pyconnection.cpp (hunk 2 above)
Comment (Tishj) [reply tools/pythonpkg/src/pyconnection.cpp:438]: I tried that, and I'm pretty sure that also reported true for False, hence this ugly workaround

File: tools/pythonpkg/src/pyconnection.cpp (hunk 3 above)
Comment (Tishj) [reply tools/pythonpkg/src/pyconnection.cpp:418]: To copy the API of Pandas' read_csv function, if I didn't use kwargs they would have to be specified in the right order, no?

Comment (Mause) [reply tools/pythonpkg/src/pyconnection.cpp:418]: You can provide keyword arguments in any order you like, it's only positional arguments that need to be provided in the right order (hence why they are positional)

File: tools/pythonpkg/src/pyconnection.cpp (hunk 2 above)
Comment (Mause) [reply tools/pythonpkg/src/pyconnection.cpp:438]: I used it here, and it seemed to work okay? https://github.com/duckdb/duckdb/blob/b793c7459a63570ee53d2b683276dde19667a90e/tools/pythonpkg/src/pyfilesystem.cpp#L73

File: tools/pythonpkg/src/pyconnection.cpp (hunk 3 above)
Comment (Tishj) [reply tools/pythonpkg/src/pyconnection.cpp:418]: I'm not entirely sure what you have in mind to replace this

The keywords arguments are optional and don't have a default value

File: tools/pythonpkg/duckdb_python.cpp (hunk 4 above)
Comment (Tishj) [reply tools/pythonpkg/duckdb_python.cpp:107]: Ah nice, found how to do it with pybind11https://pybind11.readthedocs.io/en/stable/advanced/misc.html#generating-documentation-using-sphinx, will add it :)

File: tools/pythonpkg/src/pyconnection.cpp (hunk 3 above)
Comment (Mause) [reply tools/pythonpkg/src/pyconnection.cpp:452]: Using `None` is exactly how you denote a parameter as being optional in Python actually.

You can have a look at the doc to see how they've defined it: https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

Unfortunately I haven't had the chance to complete, as I'm currently travelling, but you can look at the `Optional` class here for an example: https://github.com/duckdb/duckdb/pull/5732/files#diff-09d39950a22a7a86ba6b9b1fa29ab796c2d4a30196a64bb924f196cd8390a752

File: tools/pythonpkg/src/pyconnection.cpp [hunk 5]
```diff
@@ -130,6 +133,42 @@ static void InitializeConnectionMethods(py::class_<DuckDBPyConnection, shared_pt
...
+                Read a CSV file
+                
+                :param name:
+                    The name used to identify the CSV file
+                :type name: ``str``
+                :param \**kwargs:
+                    See below
+                
+                :Keyword Arguments:
+                    * *header* (``bool``) --
```
Comment (Mause) [inline tools/pythonpkg/src/pyconnection.cpp:146]: Is there a reason you haven't formatted these using the Sphinx syntax, as you've done with the arguments above?

Comment (Tishj) [reply tools/pythonpkg/src/pyconnection.cpp:146]: https://stackoverflow.com/a/27724915

I'll try to write it using None though

File: tools/pythonpkg/tests/fast/api/test_read_csv.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+	def test_header_true(self, duckdb_cursor):
+		rel = duckdb_cursor.read_csv(TestFile('category.csv'), header=True)
+		res = rel.fetchone()
+		print(res)
+		assert res == (1, 'Action', datetime.datetime(2006, 2, 15, 4, 46, 27))
+
+	# uncomment after issue #6011 is fixed
+	#def test_header_false(self, duckdb_cursor):
+	#	rel = duckdb_cursor.read_csv(TestFile('category.csv'), header=False)
```
Comment (pdet) [inline tools/pythonpkg/tests/fast/api/test_read_csv.py:72]: maybe a 
`@pytest.mark.skip(reason="...")`

File: tools/pythonpkg/tests/fast/api/test_read_csv.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import pandas
+import pytest
+import duckdb
+
+def TestFile(name):
+	import os
+	filename = os.path.join(os.path.dirname(os.path.realpath(__file__)),'..','data',name)
+	return filename
+
+class TestReadCSV(object):
```
Comment (pdet) [inline tools/pythonpkg/tests/fast/api/test_read_csv.py:12]: Can we also add a test with the parallel_csv_reader?

**Comment by Tishj** [comment]
I have a segfault that I really can't wrap my head around

Test:
```
	def test_encoding(self, duckdb_cursor):
		with pytest.raises(duckdb.BinderException, match="Copy is only supported for UTF-8 encoded files, ENCODING 'UTF-8'"):
			rel = duckdb_cursor.read_csv(TestFile('quote_escape.csv'), encoding=";")
```

Result:
```
  Fatal Python error: Segmentation fault
  
  Current thread 0x00007f84cd397740 (most recent call first):
    File "/project/tests/fast/api/test_read_csv.py", line 105 in test_encoding
```

The only thing I'm doing with `encoding` is checking it for "utf-8" and throwing an error if it's not, nothing else get's done for it
```c++
	if (!py::none().is(encoding)) {
		if (!py::isinstance<py::str>(encoding)) {
			throw InvalidInputException("read_csv only accepts 'encoding' as a string");
		}
		string encoding = StringUtil::Lower(py::str(encoding));
		if (encoding != "utf8" && encoding != "utf-8") {
			throw BinderException("Copy is only supported for UTF-8 encoded files, ENCODING 'UTF-8'");
		}
	}
```

File: tools/pythonpkg/tests/fast/api/test_read_csv.py (hunk 7 above)
Comment (Mytherin) [reply tools/pythonpkg/tests/fast/api/test_read_csv.py:12]: > So maybe we should just upgrade it to a csv reader option?

I think that is a good idea

Comment (Tishj) [reply tools/pythonpkg/tests/fast/api/test_read_csv.py:12]: Removing the DBConfig setting as well, or not yet?

**Comment by Tishj** [comment]
I am kind of confused by this giant regression
```
pandas_load_lineitem
Old timing: 3.6648285388946533
New timing: 9.347676873207092
```
It seems consistent, but looking at `regression_test_python.py::run_dataload` I can't discern how these changes would affect it

**Review by Mytherin — APPROVED** [review]
Thanks for the PR! LGTM

**Comment by Tishj** [comment]
Thanks, I also have a branch thats nearly ready for a PR for the `read_parquet`, `to_csv`, `to_parquet` + `write_parquet` (alias) methods

And the cleanup we talked about to the read_csv_relation internals

**Pull request merged**